File: jas_ocaml/lib/drawing_tool.ml
(** Drawing tools: line, rect, polygon. *)

(* ------------------------------------------------------------------ *)
(* Drawing tool base                                                   *)
(* ------------------------------------------------------------------ *)

class virtual drawing_tool_base = object (_self)
  val mutable drag_start : (float * float) option = None
  val mutable drag_end : (float * float) option = None

  method virtual private create_element : float -> float -> float -> float -> Element.element option
  method virtual private draw_preview : Cairo.context -> float -> float -> float -> float -> unit

  method on_press (_ctx : Canvas_tool.tool_context) x y ~(shift : bool) ~(alt : bool) =
    ignore (shift, alt);
    drag_start <- Some (x, y);
    drag_end <- Some (x, y)

  method on_move (ctx : Canvas_tool.tool_context) x y ~shift ~(dragging : bool) =
    ignore dragging;
    match drag_start with
    | Some (sx, sy) ->
      let (cx, cy) = if shift then Canvas_tool.constrain_angle sx sy x y else (x, y) in
      drag_end <- Some (cx, cy);
      ctx.request_update ()
    | None -> ()

  method on_release (ctx : Canvas_tool.tool_context) x y ~shift ~(alt : bool) =
    ignore alt;
    match drag_start with
    | None -> ()
    | Some (sx, sy) ->
      let (ex, ey) = if shift then Canvas_tool.constrain_angle sx sy x y else (x, y) in
      drag_start <- None;
      drag_end <- None;
      (match _self#create_element sx sy ex ey with
       | Some elem -> ctx.controller#add_element elem
       | None -> ())

  method on_double_click (_ctx : Canvas_tool.tool_context) (_x : float) (_y : float) = ()
  method on_key (_ctx : Canvas_tool.tool_context) (_key : int) = false
  method activate (_ctx : Canvas_tool.tool_context) = ()
  method deactivate (_ctx : Canvas_tool.tool_context) = ()

  method draw_overlay (_ctx : Canvas_tool.tool_context) cr =
    match drag_start, drag_end with
    | Some (sx, sy), Some (ex, ey) ->
      Cairo.set_source_rgba cr 0.4 0.4 0.4 1.0;
      Cairo.set_line_width cr 1.0;
      Cairo.set_dash cr [| 4.0; 4.0 |];
      _self#draw_preview cr sx sy ex ey;
      Cairo.stroke cr;
      Cairo.set_dash cr [||]
    | _ -> ()
end

(* ------------------------------------------------------------------ *)
(* Line tool                                                           *)
(* ------------------------------------------------------------------ *)

class line_tool = object
  inherit drawing_tool_base

  method private create_element sx sy ex ey =
    Some (Element.Line {
      x1 = sx; y1 = sy; x2 = ex; y2 = ey;
      stroke = Canvas_tool.default_stroke;
      opacity = 1.0; transform = None;
    })

  method private draw_preview cr sx sy ex ey =
    Cairo.move_to cr sx sy;
    Cairo.line_to cr ex ey
end

(* ------------------------------------------------------------------ *)
(* Rect tool                                                           *)
(* ------------------------------------------------------------------ *)

class rect_tool = object
  inherit drawing_tool_base

  method private create_element sx sy ex ey =
    Some (Element.Rect {
      x = min sx ex; y = min sy ey;
      width = abs_float (ex -. sx); height = abs_float (ey -. sy);
      rx = 0.0; ry = 0.0;
      fill = None; stroke = Canvas_tool.default_stroke;
      opacity = 1.0; transform = None;
    })

  method private draw_preview cr sx sy ex ey =
    let rx = min sx ex and ry = min sy ey in
    let rw = abs_float (ex -. sx) and rh = abs_float (ey -. sy) in
    Cairo.rectangle cr rx ry ~w:rw ~h:rh
end

(* ------------------------------------------------------------------ *)
(* Polygon tool                                                        *)
(* ------------------------------------------------------------------ *)

class polygon_tool = object
  inherit drawing_tool_base

  method private create_element sx sy ex ey =
    let pts = Canvas_tool.regular_polygon_points sx sy ex ey Canvas_tool.polygon_sides in
    Some (Element.Polygon {
      points = pts;
      fill = None; stroke = Canvas_tool.default_stroke;
      opacity = 1.0; transform = None;
    })

  method private draw_preview cr sx sy ex ey =
    let pts = Canvas_tool.regular_polygon_points sx sy ex ey Canvas_tool.polygon_sides in
    match pts with
    | (fx, fy) :: rest ->
      Cairo.move_to cr fx fy;
      List.iter (fun (px, py) -> Cairo.line_to cr px py) rest;
      Cairo.Path.close cr
    | [] -> ()
end

File: jas/tools/drawing_tool.py
"""Drawing tool base class shared by line / rect / rounded_rect / polygon / star.

The individual drawing tools live in their own per-tool files
(`line_tool.py`, `rect_tool.py`, etc.) and import `DrawingToolBase`
from here. This file holds only the base class plus the angle-snap
helper used by every drawing tool.
"""

from __future__ import annotations

import math
from typing import TYPE_CHECKING

from tools.tool import CanvasTool, ToolContext

if TYPE_CHECKING:
    from PySide6.QtGui import QPainter


def _constrain_angle(sx: float, sy: float, ex: float, ey: float) -> tuple[float, float]:
    dx = ex - sx
    dy = ey - sy
    dist = math.hypot(dx, dy)
    if dist == 0:
        return (ex, ey)
    angle = math.atan2(dy, dx)
    snapped = round(angle / (math.pi / 4)) * (math.pi / 4)
    return (sx + dist * math.cos(snapped), sy + dist * math.sin(snapped))


class DrawingToolBase(CanvasTool):
    """Base for press-drag-release drawing tools."""

    def __init__(self):
        self._drag_start: tuple[float, float] | None = None
        self._drag_end: tuple[float, float] | None = None

    def on_press(self, ctx: ToolContext, x: float, y: float,
                 shift: bool = False, alt: bool = False) -> None:
        ctx.snapshot()
        self._drag_start = (x, y)
        self._drag_end = (x, y)

    def on_move(self, ctx: ToolContext, x: float, y: float,
                shift: bool = False, dragging: bool = False) -> None:
        if self._drag_start is not None:
            if shift:
                x, y = _constrain_angle(*self._drag_start, x, y)
            self._drag_end = (x, y)
            ctx.request_update()

    def on_release(self, ctx: ToolContext, x: float, y: float,
                   shift: bool = False, alt: bool = False) -> None:
        if self._drag_start is None:
            return
        sx, sy = self._drag_start
        if shift:
            x, y = _constrain_angle(sx, sy, x, y)
        self._drag_start = None
        self._drag_end = None
        elem = self._create_element(sx, sy, x, y)
        if elem is not None:
            ctx.controller.add_element(elem)

    def _create_element(self, sx, sy, ex, ey):
        raise NotImplementedError

    def _draw_preview(self, painter: QPainter, sx, sy, ex, ey) -> None:
        raise NotImplementedError

    def draw_overlay(self, ctx: ToolContext, painter: QPainter) -> None:
        from PySide6.QtGui import QColor, QPen, Qt
        if self._drag_start is None or self._drag_end is None:
            return
        sx, sy = self._drag_start
        ex, ey = self._drag_end
        pen = QPen(QColor(100, 100, 100), 1.0, Qt.PenStyle.DashLine)
        painter.setPen(pen)
        from PySide6.QtGui import QBrush
        painter.setBrush(QBrush())
        self._draw_preview(painter, sx, sy, ex, ey)
